src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+			rteToSubqueryConverterReference->rangeTableEntry,
+			rteToSubqueryConverterReference->restrictionList,
+			rteToSubqueryConverterReference->
+			requiredAttributeNumbers,
+			context);
+		RemoveFromRTEToSubqueryConverterContext(rteToSubqueryConverterContext,
+												rteToSubqueryConverterReference->
+												rangeTableEntry->relid);
+		rteToSubqueryConverterReference =
+			GetNextRTEToConvertToSubquery(query->jointree, rteToSubqueryConverterContext,
+										  context->plannerRestrictionContext,
+										  resultRelationId);
+	}
+}
+
+
+/*
+ * GetResultRelationId gets the result relation id from query
+ * if it exists.
+ */
+static Oid
+GetResultRelationId(Query *query)
+{
+	RangeTblEntry *resultRelation = ExtractResultRelationRTE(query);
+	Oid resultRelationId = InvalidOid;
+	if (resultRelation)
+	{
+		resultRelationId = resultRelation->relid;
+	}
+	return resultRelationId;
+}
+
+
+/*
+ * GetRangeTableEntriesFromJoinTree gets the range table entries that are
+ * on the given join tree.
+ */
+static void
+GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+								 List **joinRangeTableEntries)
+{
+	if (joinNode == NULL)
+	{
+		return;
+	}
+	else if (IsA(joinNode, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) joinNode;
+		Node *fromElement;
+
+		foreach_ptr(fromElement, fromExpr->fromlist)
+		{
+			GetRangeTableEntriesFromJoinTree(fromElement, rangeTableList,
+											 joinRangeTableEntries);
+		}
+	}
+	else if (IsA(joinNode, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		GetRangeTableEntriesFromJoinTree(joinExpr->larg, rangeTableList,
+										 joinRangeTableEntries);
+		GetRangeTableEntriesFromJoinTree(joinExpr->rarg, rangeTableList,
+										 joinRangeTableEntries);
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		int rangeTableIndex = ((RangeTblRef *) joinNode)->rtindex;
+		RangeTblEntry *rte = rt_fetch(rangeTableIndex, rangeTableList);
+		*joinRangeTableEntries = lappend(*joinRangeTableEntries, rte);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * GetNextRTEToConvertToSubquery returns the range table entry
+ * which should be converted to a subquery. It considers the local join policy
+ * and result relation.
+ */
+static RTEToSubqueryConverterReference *
+GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+							  RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+							  PlannerRestrictionContext *plannerRestrictionContext, Oid
+							  resultRelationId)
+{
+	RTEToSubqueryConverterReference *localRTECandidate = NULL;
+	RTEToSubqueryConverterReference *distributedRTECandidate = NULL;
+	if (!FillLocalAndDistributedRTECandidates(rteToSubqueryConverterContext,
+											  &localRTECandidate,
+											  &distributedRTECandidate))
+	{
+		return NULL;
+	}
+
+	if (OidIsValid(resultRelationId))
+	{
+		if (resultRelationId == GetRTEToSubqueryConverterReferenceRelId(
+				localRTECandidate))
+		{
+			return distributedRTECandidate;
```

Comment:
This seems quite messy/confusing, could we take the resultRelationId out of the list of candidates in advance?

Reply:
umm, I guess that would work, which is also simpler.